src/miner_lora_throttle.erl

```diff
@@ -0,0 +1,224 @@
+
+-type regulatory_model() :: {dwell | duty, Limit :: number(), Period :: number()}.
+
+-opaque handle() :: {regulatory_model(), list(#sent_packet{})}.
+
+%% Maximum time allowable time on air.
+-define(MAX_TIME_ON_AIR_MS, 400).
+
+-spec model(region()) -> regulatory_model().
+model(Region) ->
+    case Region of
+        %%                   Limit  Period
+        %%                     (%)
+        'AS923' -> {'duty',   0.01, 3600000};
+        'AU915' -> {'duty',   0.01, 3600000};
+        'CN470' -> {'duty',   0.01, 3600000};
+        'CN779' -> {'duty',   0.01, 3600000};
+        'EU433' -> {'duty',   0.01, 3600000};
+        'EU868' -> {'duty',   0.01, 3600000};
+        'IN865' -> {'duty',   0.01, 3600000};
+        'KR920' -> {'duty',   0.01, 3600000};
+        %%                      ms  Period
+        'US915' -> {'dwell',   400,   20000}
+    end.
+
+%% Updates Handle with time-on-air information.
+%%
+%% This function does not send/transmit itself.
+-spec track_sent(
+    Handle :: handle(),
+    SentAt :: number(),
+    Frequency :: number(),
+    Bandwidth :: number(),
+    SpreadingFactor :: integer(),
+    CodeRate :: integer(),
+    PreambleSymbols :: integer(),
+    ExplicitHeader :: boolean(),
+    PayloadLen :: integer()
+) ->
+    handle().
+track_sent(
+    Handle,
+    SentAt,
+    Frequency,
+    Bandwidth,
+    SpreadingFactor,
+    CodeRate,
+    PreambleSymbols,
+    ExplicitHeader,
+    PayloadLen
+) ->
+    TimeOnAir = time_on_air(
+        Bandwidth,
+        SpreadingFactor,
+        CodeRate,
+        PreambleSymbols,
+        ExplicitHeader,
+        PayloadLen
+    ),
+    track_sent(Handle, SentAt, Frequency, TimeOnAir).
+
+-spec track_sent(handle(), number(), number(), number()) -> handle().
+track_sent({Region, SentPackets}, SentAt, Frequency, TimeOnAir) ->
+    NewSent = #sent_packet{
+        frequency = Frequency,
+        sent_at = SentAt,
+        time_on_air = TimeOnAir
+    },
+    {Region, trim_sent(Region, [NewSent | SentPackets])}.
+
+-spec trim_sent(regulatory_model(), list(#sent_packet{})) -> list(#sent_packet{}).
+trim_sent(Model, SentPackets = [NewSent, LastSent | _])
+        when NewSent#sent_packet.sent_at < LastSent#sent_packet.sent_at ->
+    trim_sent(Model, lists:sort(fun (A, B) -> A > B end, SentPackets));
+trim_sent({_, _, Period}, SentPackets = [H | _]) ->
+    CutoffTime = H#sent_packet.sent_at - Period,
+    Pred = fun (Sent) -> Sent#sent_packet.sent_at > CutoffTime end,
+    lists:takewhile(Pred, SentPackets).
+
+%% @doc Based on previously sent packets, returns a boolean value if
+%% it is legal to send on Frequency at time Now.
+%%
+%%
+-spec can_send(
+    Handle :: handle(),
+    AtTime :: number(),
+    Frequency :: integer(),
+    TimeOnAir :: number()
+) ->
+    boolean().
+can_send(_Handle, _AtTime, _Frequency, TimeOnAir) when TimeOnAir > ?MAX_TIME_ON_AIR_MS ->
+    %% TODO: check that all regions have do in fact have the same
```

Comment:
Is there a way to resolve this?

Reply:
The rules are a bit nebulous, and it seems like everyone interprets them differently.  From what I can tell, it is resolved, and this will not allow for too-long-on-air. I do not _think_ it will artificially restrict time on air to what's allowed, but not 100% certain.